src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/base/SalesforceBaseBatchSource.java

```diff
@@ -0,0 +1,148 @@
+import io.cdap.cdap.etl.api.batch.BatchSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Base class for Salesforce Batch Source plugins. Contains data transformation logic.
+ *
+ * @param <T> input data key value
+ */
+public abstract class SalesforceBaseBatchSource<T> extends BatchSource<T, Map<String, String>, StructuredRecord> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SalesforceBaseBatchSource.class);
+
+  private static final String ERROR_SCHEMA_BODY_PROPERTY = "body";
+
+  private static final Schema ERROR_SCHEMA = Schema.recordOf("error",
+    Schema.Field.of(ERROR_SCHEMA_BODY_PROPERTY, Schema.of(Schema.Type.STRING)));
+
+  @Override
+  public void transform(KeyValue<T, Map<String, String>> input,
+                        Emitter<StructuredRecord> emitter) {
+    Map<String, String> record = input.getValue();
+    try {
+      Schema schema = getRecordSchema(input.getKey());
+      StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+
+      for (Map.Entry<String, String> entry : record.entrySet()) {
+        String fieldName = entry.getKey();
+        String value = entry.getValue();
+
+        Schema.Field field = schema.getField(fieldName, true);
+
+        if (field == null) {
+          continue; // this field is not in schema
+        }
+
+        builder.set(field.getName(), convertValue(value, field));
+      }
+
+      emitter.emit(builder.build());
+    } catch (Exception ex) {
+      switch (getConfig().getErrorHandling()) {
+        case SKIP:
+          LOG.warn("Cannot process Salesforce row '{}', skipping it.", record, ex);
+          break;
+        case SEND:
+          StructuredRecord.Builder builder = StructuredRecord.builder(ERROR_SCHEMA);
+          builder.set(ERROR_SCHEMA_BODY_PROPERTY, record);
+          emitter.emitError(new InvalidEntry<>(400, ex.getMessage(), builder.build()));
+          break;
+        case STOP:
+          throw ex;
+        default:
+          throw new UnexpectedFormatException(
+            String.format("Unknown error handling strategy '%s'", getConfig().getErrorHandling()));
+      }
+    }
+  }
+
+  protected abstract SalesforceBaseSourceConfig getConfig();
+
+  protected abstract Schema getRecordSchema(T key);
+
+  private Object convertValue(String value, Schema.Field field) {
+    Schema fieldSchema = field.getSchema();
+
+    if (fieldSchema.isNullable()) {
+      fieldSchema = fieldSchema.getNonNullable();
+    }
+
+    Schema.Type fieldSchemaType = fieldSchema.getType();
+
+    // empty string is considered null in csv
+    if (Strings.isNullOrEmpty(value)) {
+      return null;
+    }
+
+    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+    if (fieldSchema.getLogicalType() != null) {
+      switch (logicalType) {
+        case DATE:
```

Comment:
I've seen this switch case structure around a few times now in various code reviews. Does this refactor capture all usages of it? If not we should extract this to a util/helper class